crates/ruma-client-api/src/discovery/discover_support.rs

```diff
@@ -0,0 +1,117 @@
+//! `GET /.well-known/matrix/support` ([spec])
+//!
+//! [spec]: https://spec.matrix.org/latest/client-server-api/#getwell-knownmatrixsupport
+//!
+//! Get server admin contact and support page of a homeserver's domain.
+
+use ruma_common::{
+    api::{request, response, Metadata},
+    metadata,
+    serde::StringEnum,
+    OwnedUserId,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::PrivOwnedStr;
+
+const METADATA: Metadata = metadata! {
+    method: GET,
+    rate_limited: false,
+    authentication: None,
+    history: {
+        1.10 => "/.well-known/matrix/support",
+    }
+};
+
+/// Request type for the `discover_support` endpoint.
+#[request(error = crate::Error)]
+#[derive(Default)]
+pub struct Request {}
+
+/// Response type for the `discover_support` endpoint.
+#[response(error = crate::Error)]
+pub struct Response {
+    /// Ways to contact the server administrator.
+    ///
+    /// At least one of `contacts` or `support_page` is required. If only `contacts` is set, it
+    /// must contain at least one item.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub contacts: Vec<Contact>,
+
+    /// The URL of a page to give users help specific to the homeserver, like extra
+    /// login/registration steps.
+    ///
+    /// At least one of `contacts` or `support_page` is required.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub support_page: Option<String>,
+}
+
+impl Request {
+    /// Creates an empty `Request`.
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Response {
+    /// Creates a new `Response` with the given contacts.
+    pub fn with_contacts(contacts: Vec<Contact>) -> Self {
+        Self { contacts, support_page: None }
+    }
+
+    /// Creates a new `Response` with the given support page.
+    pub fn with_support_page(support_page: String) -> Self {
+        Self { contacts: Vec::new(), support_page: Some(support_page) }
+    }
+}
+
+/// A way to contact the server administrator.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Contact {
+    /// An informal description of what the contact methods are used for.
+    pub role: ContactRole,
+
+    /// An email address to reach the administrator.
+    ///
+    /// At least one of `matrix_id` or `email_address` is required.
```

Comment:
Why not use an enum like elsewhere? Just because it's extra effort for a not super important endpoint?

Reply:
Yes, mostly. I don't think having strong validation here brings anything.

Also I'm not sure if we already have enums with variants like `A(_), B(_), AandB(_, _)`.

Reply:
Ohh, this is at least one, not xor. Alright, let's not waste time on the extra validation then.